File: sbt-plugin/src/main/scala/play/soap/sbtplugin/FutureGenerator.scala
/*
 * Copyright (C) 2015-2019 Lightbend Inc. <https://www.lightbend.com>
 */
package play.soap.sbtplugin

import play.soap.sbtplugin.Imports.WsdlKeys.FutureApi

/**
 * A helper used by [[PlaySEIGenerator]] to generate the correct future type to wrap a Java type.
 * @param futureApi
 */
class FutureGenerator(futureApi: FutureApi) {

  /** Pregenerated type mappings */
  private val mapping: Map[String, String] = Map(
    "void"    -> futureApi.voidType,
    "boolean" -> "java.lang.Boolean",
    "byte"    -> "java.lang.Byte",
    "char"    -> "java.lang.Character",
    "double"  -> "java.lang.Double",
    "float"   -> "java.lang.Float",
    "int"     -> "java.lang.Integer",
    "long"    -> "java.lang.Long",
    "short"   -> "java.lang.Short"
  )

  /**
   * Generate the correct future type that holds the given Java type. E.g. for a Java type of
   * "void" this could be "scala.concurrent.Future<scala.Unit>" or "java.util.concurrent.CompletionStage<Void>".
   * For a Java type of "int" it could be "scala.concurrent.Future<java.lang.Integer>" or
   * "java.util.concurrent.CompletionStage<java.lang.Integer>"
   *
   * @param javaType The Java type to wrap.
   */
  def futureType(javaType: String) = {
    val elementType = mapping.getOrElse(javaType, javaType)
    futureApi.fqn + "<" + elementType + ">"
  }
}

File: plugin/src/main/java/play/soap/plugin/FutureGenerator.java
/*
 * Copyright (C) from 2022 The Play Framework Contributors <https://github.com/playframework>, 2011-2021 Lightbend Inc. <https://www.lightbend.com>
 */
package play.soap.plugin;

import java.util.HashMap;
import java.util.Map;

/**
 * A helper used by {@link PlaySEIGenerator} to generate the correct future type to wrap a Java
 * type.
 */
public class FutureGenerator {

  private final FutureApi futureApi;

  private final Map<String, String> mapping;

  public FutureGenerator(FutureApi futureApi) {
    this.futureApi = futureApi;
    this.mapping =
        new HashMap<String, String>() {
          {
            put("void", futureApi.voidType());
            put("boolean", "java.lang.Boolean");
            put("byte", "java.lang.Byte");
            put("char", "java.lang.Character");
            put("double", "java.lang.Double");
            put("float", "java.lang.Float");
            put("int", "java.lang.Integer");
            put("long", "java.lang.Long");
            put("short", "java.lang.Short");
          }
        };
  }

  /**
   * Generate the correct future type that holds the given Java type. E.g. for a Java type of "void"
   * this could be "scala.concurrent.Future {@literal <}scala.Unit{@literal >}" or
   * "java.util.concurrent.CompletionStage{@literal <}Void{@literal >}". For a Java type of "int" it
   * could be "scala.concurrent.Future{@literal <}java.lang.Integer{@literal >}" or
   * "java.util.concurrent.CompletionStage{@literal <}java.lang.Integer{@literal >}"
   *
   * @param javaType The Java type to wrap.
   */
  public String futureType(String javaType) {
    String type = mapping.get(javaType);
    String elementType = type != null ? type : javaType;
    return futureApi.fqn() + "<" + elementType + ">";
  }
}
